src/utils/pnetcdf_version/pnetcdf_version.1
.\" $Header$
.TH pnetcdf_version 1 "PnetCDF PNETCDF_RELEASE_VERSION" "Last modified: GIT_SRC_LAST_COMMIT" "PnetCDF utilities"
.SH NAME
pnetcdf_version \- print the version information of PnetCDF library
.SH SYNOPSIS
.ft B
.HP
pnetcdf_version
.nh
\%[-v]
\%[-d]
\%[-c]
\%[-b]
\%[-l]
\%[-h]
.hy
.ft
.SH DESCRIPTION
\fBpnetcdf_version\fP prints the version information of PnetCDF library and
the configure command line used to build the library.

If no argument is given, all information is printed.
.SH OPTIONS
.IP "\fB-v\fP"
Version number of this PnetCDF release.
.IP "\fB-d\fP"
Release date.
.IP "\fB-c\fP"
Configure command-line arguments used to build this PnetCDF
.IP "\fB-b\fP"
MPI compilers used to build this PnetCDF library
.IP "\fB-l\fP"
LMOD PrgEnv module loaded, if available
.IP "\fB-h\fP"
Print the available command-line options of pnetcdf_version

.SH EXAMPLES
.LP
Print all information about the PnetCDF library by running the command with no options.

% pnetcdf_version
.sp
.nf
PnetCDF Version:    	PNETCDF_RELEASE_VERSION
PnetCDF Release date:	PNETCDF_RELEASE_DATE
PnetCDF configure: 	    --with-mpi=/usr/local/bin

MPICC:  /opt/cray/pe/craype/2.7.35/bin/cc
        CFLAGS: -O2
        MPI standard version: 4.0
        MPI compiler vendor: Cray MPICH 9.0.1.498
        Base compiler: /usr/bin/gcc-14
        Base compiler version: gcc-14 (SUSE Linux) 14.3.0

MPICXX: /opt/cray/pe/craype/2.7.35/bin/CC
        CXXFLAGS: -O2
        Base compiler: /usr/bin/g++-14
        Base compiler version: g++-14 (SUSE Linux) 14.3.0

MPIF77: /opt/cray/pe/craype/2.7.35/bin/ftn
        FFLAGS: -O2
        Base compiler: /opt/cray/pe/gcc-native/14/bin/gfortran
        Base compiler version: GNU Fortran (SUSE Linux) 14.3.0

MPIF90: /opt/cray/pe/craype/2.7.35/bin/ftn
        FCFLAGS: -O2
        Base compiler: /opt/cray/pe/gcc-native/14/bin/gfortran
        Base compiler version: GNU Fortran (SUSE Linux) 14.3.0

LMOD module loaded:
        LMOD_VERSION                    8.7.60
        LMOD_FAMILY_PRGENV              PrgEnv-gnu
        LMOD_FAMILY_PRGENV_VERSION      8.6.0
        LMOD_FAMILY_COMPILER            gcc-native
        LMOD_FAMILY_COMPILER_VERSION    14
        LMOD_FAMILY_MPI                 cray-mpich
        LMOD_FAMILY_MPI_VERSION         9.0.1
.fi

.SH "SEE ALSO"
.LP
.BR pnetcdf (3)
.SH DATE
PNETCDF_RELEASE_DATE
.LP




NEWS
Please see PnetCDF project web sites:
https://parallel-netcdf.github.io
http://cucis.ece.northwestern.edu/projects/PnetCDF


src/utils/pnetcdf_version/pnetcdf_version.c
/*
 *  Copyright (C) 2014, Northwestern University and Argonne National Laboratory
 *  See COPYRIGHT notice in top-level directory.
 */
/* $Id$ */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h> /* getopt() */

typedef enum {
    Version_number = 0,
    Date           = 1,
    Patches        = 2,
    Configure_args = 3,
    Compilers      = 4,
    Lmod           = 5,
    LastField
} fields;

/*
 *   pnetcdf_version - Report on the PnetCDF version
 */

/*---< usage() >-------------------------------------------------------------*/
static void usage(char *argv0) {
    char *help =
        "Usage: %s [switches]\n"
        "       -v  : version number\n"
        "       -d  : release date\n"
        "       -c  : configure arguments used to build PnetCDF\n"
        "       -b  : MPI compilers used\n"
        "       -l  : LMOD PrgEnv module loaded, if available\n"
        "       -h  : print this help (available command-line options)\n";
    fprintf(stderr, help, argv0);
    exit(-1);
}

/*----< main() >-------------------------------------------------------------*/
int main( int argc, char *argv[] )
{
    int i, opt, flags[10];

    if (argc <= 1) {
        /* Show all values */
        for (i=0; i<LastField; i++) flags[i] = 1;
    }
    else {
        /* Show only requested values */
        for (i=0; i<LastField; i++) flags[i] = 0;
    }

    while ( (opt=getopt(argc,argv,"vdcblh"))!= EOF) {
        switch (opt) {
            case 'v': flags[Version_number] = 1;
                      break;
            case 'd': flags[Date] = 1;
                      break;
            case 'c': flags[Configure_args] = 1;
                      break;
            case 'b': flags[Compilers] = 1;
                      break;
            case 'l': flags[Lmod] = 1;
                      break;
            case 'h':
            default: usage(argv[0]);
                      break;
        }
    }

    /* Print out the information, one item per line */
    if (flags[Version_number])
        printf("PnetCDF Version:    \t%s\n", PNETCDF_VERSION);

    if (flags[Date])
        printf("PnetCDF Release date:\tPNETCDF_RELEASE_DATE\n");

    if (flags[Configure_args])
        printf("PnetCDF configure: \t%s\n", CONFIGURE_ARGS_CLEAN);

    printf("\n");

    if (flags[Compilers]) {
        printf("MPICC:  %s\n", MPICC);
#ifdef CFLAGS
        if (strcmp(CFLAGS, ""))
            printf("        CFLAGS: %s\n", CFLAGS);
        else
            printf("        CFLAGS:\n");
#else
        printf("        CFLAGS:\n");
#endif
        printf("        MPI standard version: %s\n", MPI_VERSION);
        printf("        MPI compiler vendor: %s %s\n", MPI_VENDOR, MPI_VENDOR_VERSION);
        printf("        Base compiler: %s\n", MPICC_BASE);
        printf("        Base compiler version: %s\n", MPICC_BASE_VERSION);
        printf("\n");

#ifdef MPICXX
        printf("MPICXX: %s\n", MPICXX);
#ifdef CXXFLAGS
        if (strcmp(CXXFLAGS, ""))
            printf("        CXXFLAGS: %s\n", CXXFLAGS);
        else
            printf("        CXXFLAGS\n");
#else
        printf("        CXXFLAGS\n");
#endif
        printf("        Base compiler: %s\n", MPICXX_BASE);
        printf("        Base compiler version: %s\n", MPICXX_BASE_VERSION);
        printf("\n");
#endif

#ifdef MPIF77
        printf("MPIF77: %s\n", MPIF77);
#ifdef FFLAGS
        if (strcmp(FFLAGS, ""))
            printf("        FFLAGS: %s\n", FFLAGS);
        else
            printf("        FFLAGS:\n");
#else
        printf("        FFLAGS:\n");
#endif
        printf("        Base compiler: %s\n", MPIF77_BASE);
        printf("        Base compiler version: %s\n", MPIF77_BASE_VERSION);
        printf("\n");
#endif

#ifdef MPIF90
        printf("MPIF90: %s\n", MPIF90);
#ifdef FCFLAGS
        if (strcmp(FCFLAGS, ""))
            printf("        FCFLAGS: %s\n", FCFLAGS);
        else
            printf("        FCFLAGS:\n");
#else
        printf("        FCFLAGS:\n");
#endif
        printf("        Base compiler: %s\n", MPIF90_BASE);
        printf("        Base compiler version: %s\n", MPIF90_BASE_VERSION);
        printf("\n");
#endif
    }

    if (flags[Lmod]) {
#ifdef LMOD_FAMILY_PRGENV
        printf("LMOD module loaded:\n");
        printf("\tLMOD_VERSION\t\t\t%s\n", LMOD_VERSION);
        printf("\tLMOD_FAMILY_PRGENV\t\t%s\n", LMOD_FAMILY_PRGENV);
        printf("\tLMOD_FAMILY_PRGENV_VERSION\t%s\n", LMOD_FAMILY_PRGENV_VERSION);
        printf("\tLMOD_FAMILY_COMPILER\t\t%s\n", LMOD_FAMILY_COMPILER);
        printf("\tLMOD_FAMILY_COMPILER_VERSION\t%s\n", LMOD_FAMILY_COMPILER_VERSION);
        printf("\tLMOD_FAMILY_MPI\t\t\t%s\n", LMOD_FAMILY_MPI);
        printf("\tLMOD_FAMILY_MPI_VERSION\t\t%s\n", LMOD_FAMILY_MPI_VERSION);
#else
        printf("LMOD PrgEnv module loaded: N/A\n");
#endif
        printf("\n");
    }

    return 0;
}

